exec/java-exec/src/main/java/org/apache/drill/exec/planner/cost/DrillRelMdDistinctRowCount.java

```diff
@@ -56,9 +73,159 @@ public Double getDistinctRowCount(Join rel, RelMetadataQuery mq,
     return getDistinctRowCount((RelNode) rel, mq, groupKey, predicate);
   }
 
-  public Double getDistinctRowCount(DrillScanRel scan, RelMetadataQuery mq,
-      ImmutableBitSet groupKey, RexNode predicate) {
-    // Consistent with the estimation of Aggregate row count in RelMdRowCount : distinctRowCount = rowCount * 10%.
-    return scan.estimateRowCount(mq) * 0.1;
+  @Override
+  public Double getDistinctRowCount(RelNode rel, RelMetadataQuery mq, ImmutableBitSet groupKey, RexNode predicate) {
+    if (rel instanceof TableScan && !DrillRelOptUtil.guessRows(rel)) {
+      return getDistinctRowCount((TableScan) rel, mq, groupKey, predicate);
+    } else if (rel instanceof SingleRel && !DrillRelOptUtil.guessRows(rel)) {
+        if (rel instanceof Window) {
+          int childFieldCount = ((Window)rel).getInput().getRowType().getFieldCount();
+          // For window aggregates delegate ndv to parent
+          for (int bit : groupKey) {
+            if (bit >= childFieldCount) {
```

Comment:
Considering that window functions don't change the NDV, it is not clear why additional check is being done here (the comment above says 'delegate ndv to parent' but this is doing it conditionally.

Reply:
I vaguely remember running into some functional test failures. This is specifically for the case when the NDV is requested for an aggregate e.g. avg, rank inside the window function. The comment says so window aggregate.